runtime/src/crowdfund.rs

```diff
@@ -0,0 +1,539 @@
+	}
+}
+
+decl_event!(
+	pub enum Event<T>
+		where
+	<T as system::Trait>::AccountId,
+	Balance = BalanceOf<T>,
+	ParaId = ParaIdOf<T>,
+	{
+		Created(FundIndex),
+		Contributed(AccountId, FundIndex, Balance),
+		Withdrew(AccountId, FundIndex, Balance),
+		Retiring(FundIndex),
+		Dissolved(FundIndex),
+		DeployDataFixed(FundIndex),
+		Onboarded(FundIndex, ParaId),
+	}
+);
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		fn deposit_event<T>() = default;
+		
+		/// Create a new crowdfunding campaign for a parachain slot deposit.
+		#[weight = TransactionWeight::Basic(100_000, 10)]
+		fn create(
+			origin,
+			#[compact] end: T::BlockNumber,
+			#[compact] cap: BalanceOf<T>,
+			#[compact] first_slot: T::BlockNumber,
+			#[compact] last_slot: T::BlockNumber
+		) {
+			let owner = ensure_signed(origin)?;
+			let deposit = T::SubmissionDeposit::get();
+
+			let imb = T::Currency::withdraw(
+				&owner,
+				deposit,
+				WithdrawReason::Transfer,
+				ExistenceRequirement::AllowDeath,
+			)?;
+
+			let index = FundCount::mutate(|c| { let r = *c; *c += 1; r });
+
+			// No fees are paid here if we need to create this account; that's why we don't just
+			// use the stock `transfer`.
+			T::Currency::resolve_creating(&Self::fund_account_id(index), imb);
+
+			<Funds<T>>::insert(index, FundInfo {
+				parachain: None,
+				owner: owner,
+				deposit: deposit,
+				raised: Zero::zero(),
+				end: end,
+				cap: cap,
+				// Ensure it's Some, so that the first contribution causes it to be inserted into
+				// `NewRaise`.
+				last_contribution: Some(Zero::zero()),
+				first_slot: first_slot,
+				last_slot: last_slot,
+				deploy_data: None,
+			});
+
+			Self::deposit_event(RawEvent::Created(index));
+
+		}
+		
+
+		
+		/// Contribute to a crowd sale. This will transfer some balance over to fund a parachain
+		/// slot. It will be withdrawable in two instances: the parachain becomes retired; or the
+		/// slot is
+		fn contribute(origin, #[compact] index: FundIndex, #[compact] value: BalanceOf<T>)
+		{
+			let who = ensure_signed(origin)?;
+
+			ensure!(value >= T::MinContribution::get(), "contribution too small");
+
+			let mut fund = Self::funds(index).ok_or("invalid fund index")?;
+			fund.raised += value;
+			ensure!(fund.raised <= fund.cap, "contributions exceed cap");
+			let now = <system::Module<T>>::block_number();
+			ensure!(fund.end > now, "contribution period ended");
+
+			T::Currency::transfer(&who, &Self::fund_account_id(index), value)?;
+
+			let id = Self::id_from_index(index);
+			let balance = who.using_encoded(|b| child::get_or_default::<BalanceOf<T>>(id.as_ref(), b));
+			let balance = balance.saturating_add(value);
+			who.using_encoded(|b| child::put(id.as_ref(), b, &balance));
+
+			if <slots::Module<T>>::is_ending(now).is_some() {
+				// Now in end period; record it
+				fund.last_contribution = Some(now);
+				if let Some(c) = fund.last_contribution {
+					if c != now {
```

Comment:
This will always be false since we just set `last_contribution` to `Some(now)`

What is the intended logic here?

Reply:
Maybe you intend to update the value _after_ the check is made.

Reply:
you're right, yes